curvefs/src/metaserver/inode_storage.cpp

```diff
@@ -171,70 +259,172 @@ MetaStatusCode InodeStorage::GetXAttr(const Key4Inode& key, XAttr *xattr) {
     return MetaStatusCode::OK;
 }
 
-MetaStatusCode InodeStorage::Delete(const Key4Inode& key) {
+MetaStatusCode InodeStorage::ForceDelete(const Key4Inode& key) {
     WriteLockGuard lg(rwLock_);
     std::string skey = conv_.SerializeToString(key);
-    Status s = kvStorage_->HDel(table4Inode_, skey);
-    if (s.ok()) {
+    std::shared_ptr<storage::StorageTransaction> txn = nullptr;
+    Status s;
+    const char* step = "Begin transaction";
+    do {
+        txn = kvStorage_->BeginTransaction();
+        if (txn == nullptr) {
+            break;
+        }
+        s = txn->HDel(table4Inode_, skey);
+        if (!s.ok()) {
+            step = "Delete inode from transaction";
+            break;
+        }
         // NOTE: for rocksdb storage, it will never check whether
         // the key exist in delete(), so if the client delete the
-        // unexist inode in some anbormal cases, it will cause the
+        // non-exist inode in some abnormal cases, it will cause the
+        // nInode less then the real value.
+        if (nInode_ > 0) {
+            s = SetInodeCount(txn.get(), nInode_ - 1);
+            if (!s.ok()) {
+                step = "Insert inode count to transaction";
+                break;
+            }
+        }
+        s = txn->Commit();
+        if (!s.ok()) {
+            step = "Delete inode";
+            break;
+        }
+        // NOTE: for rocksdb storage, it will never check whether
+        // the key exist in delete(), so if the client delete the
+        // non-exist inode in some abnormal cases, it will cause the
         // nInode less then the real value.
         if (nInode_ > 0) {
             nInode_--;
         }
         return MetaStatusCode::OK;
+    }while (false);
+    if (txn != nullptr && !txn->Rollback().ok()) {
+        LOG(ERROR) << "Rollback delete inode transaction failed, status = "
+                   << s.ToString();
     }
+    LOG(ERROR) << step << " failed, status = " << s.ToString();
+    return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+}
 
+MetaStatusCode InodeStorage::Delete(const Key4Inode& key,
+                                    int64_t logIndex) {
+    WriteLockGuard lg(rwLock_);
```

Comment:
Keeping two functions would be clearer, but there would be some problems with maintainability, as it is possible to need to modify the same code within both functions